v0/destinations/hs/util.js

```diff
@@ -1,4 +1,120 @@
-const { send } = require("../../../adapters/network");
+const get = require("get-value");
+const { send, httpGET, httpPOST } = require("../../../adapters/network");
+const {
+  getFieldValueFromMessage,
+  constructPayload,
+  CustomError,
+  isEmpty
+} = require("../../util");
+const { CONTACT_PROPERTIES, IDENTIFY_CRM_SEARCH_CONTACT } = require("./config");
+
+const formatKey = key => {
+  // lowercase and replace spaces and . with _
+  let modifiedKey = key.toLowerCase();
+  modifiedKey = modifiedKey.replace(/\s+/g, "_");
+  modifiedKey = modifiedKey.replace(/\./g, "_");
+  return modifiedKey;
+};
+
+const getTraits = message => {
+  // get from traits or properties
+  let traits = getFieldValueFromMessage(message, "traits");
+  if (!traits || !Object.keys(traits).length) {
+    traits = message.properties;
+  }
+  return traits;
+};
+
+const getProperties = async destination => {
+  let hubSpotPropertyMap = {};
+  let res;
+  const { Config } = destination;
+
+  // choosing API Type
+  if (Config.authorizationType === "newPrivateAppApi") {
+    // Private Apps
+    const requestOptions = {
+      headers: {
+        "Content-Type": "application/json",
+        Authorization: `Bearer ${Config.accessToken}`
+      }
+    };
+    res = await httpGET(CONTACT_PROPERTIES, requestOptions);
+  } else {
+    // API Key
+    const url = `${CONTACT_PROPERTIES}?hapikey=${Config.apiKey}`;
+    res = await httpGET(url);
+  }
+
+  if (res.success === false) {
+    // check if exists err.response && err.response.status else 500
+    const error = res.response;
+    if (error.response) {
+      throw new CustomError(
+        JSON.stringify(error.response.data) ||
+          JSON.stringify(error.response.statusText) ||
+          "Failed to get hubspot properties",
+        error.response.status || 500
+      );
+    }
+    throw new CustomError(
+      "Failed to get hubspot properties : invalid response",
+      500
+    );
+  }
+
+  const propertyMap = {};
+  res.response.data.forEach(element => {
+    propertyMap[element.name] = element.type;
+  });
+  hubSpotPropertyMap = propertyMap;
+  return hubSpotPropertyMap;
+};
+
+const getTransformedJSON = async (
+  message,
+  mappingJson,
+  destination,
+  propertyMap
+) => {
+  let rawPayload = {};
+  const traits = getTraits(message);
+
+  if (traits) {
+    const traitsKeys = Object.keys(traits);
+    if (!propertyMap) {
+      // fetch HS properties
+      // eslint-disable-next-line no-param-reassign
+      propertyMap = await getProperties(destination);
+    }
+
+    rawPayload = constructPayload(message, mappingJson);
+
+    // if provided data in traits matches with the
+    // property created in HS then add those in final payload
```

Comment:
- make the comment more descriptive, as per my understanding,
    
    > if there is any property in hubspot, that has not already been mapped using mapping json but exists in the traits, we will include those traits to the final payload